dhcpv6/option_userclass.go

```diff
@@ -23,29 +25,46 @@ func (op *OptUserClass) ToBytes() []byte {
 	buf := make([]byte, 6)
 	binary.BigEndian.PutUint16(buf[0:2], uint16(OPTION_USER_CLASS))
 	binary.BigEndian.PutUint16(buf[2:4], uint16(op.Length()))
-	// user-class-data has an internal data length field too..
-	binary.BigEndian.PutUint16(buf[4:6], uint16(len(op.UserClass)))
-	buf = append(buf, op.UserClass...)
+	for _, uc := range op.UserClasses {
+		binary.BigEndian.PutUint16(buf[4:6], uint16(len(uc)))
+		buf = append(buf, uc...)
+	}
 	return buf
 }
 
 // Length returns the option length
 func (op *OptUserClass) Length() int {
-	return 2 + len(op.UserClass)
+	ret := 0
+	for _, uc := range op.UserClasses {
+		ret += 2 + len(uc)
+	}
+	return ret
 }
 
 func (op *OptUserClass) String() string {
-	return fmt.Sprintf("OptUserClass{userclass=%s}", string(op.UserClass))
+	ucStrings := make([]string, 0)
+	for _, uc := range op.UserClasses {
+		ucStrings = append(ucStrings, string(uc))
+	}
+	return fmt.Sprintf("OptUserClass{userclass=[%s]}", strings.Join(ucStrings, ", "))
 }
 
 // ParseOptUserClass builds an OptUserClass structure from a sequence of
 // bytes. The input data does not include option code and length bytes.
 func ParseOptUserClass(data []byte) (*OptUserClass, error) {
 	opt := OptUserClass{}
-	dataLen := int(binary.BigEndian.Uint16(data[:2]))
-	if dataLen != len(data)-2 {
-		return nil, fmt.Errorf("ParseOptUserClass: declared data length does not match actual length: %d != %d", dataLen, len(data)-2)
+	for {
+		if len(data) == 0 {
+			break
+		}
+		if len(data) < 2 {
+			return nil, errors.New("ParseOptUserClass: short data: missing length field")
+		}
+		ucLen := int(binary.BigEndian.Uint16(data[2:]))
+        if len(data) < ucLen +2 {
+			return nil, fmt.Errorf("ParseOptUserClass: short data: less than %d bytes", ucLen +2)
+        }
+		opt.UserClasses = append(opt.UserClasses, data[2:ucLen])
```

Comment:
Isn't it missing something like `data = data[2+ucLen:]` to avoid an endless loop?

Reply:
...yes, it is. I had it in mind while writing the code and forgot it :( good catch!